### PR TITLE
Fix panic in ethstats when block state is unavailable

### PR DESCRIPTION
**Pull request by nategraf**
### Description

Under certain conditions, the `ethstats` reporting code can cause a panic due to the state of a
given block not being available to calculate some stats.

#### Panic trace

```text
2021-12-15T21:38:28.277188676Z INFO [12-15|21:38:28.276] Imported new chain segment               blocks=361 txs=3456 mgas=271.421 elapsed=4.697s      mgasps=57.779 number=9,190,047 hash=91e3a9..4a8f1b age=49m54s   dirty=33.55MiB
2021-12-15T21:38:28.336407786Z INFO [12-15|21:38:28.336] Unindexed transactions                   blocks=2030 txs=1588 tail=6,840,048 elapsed=59.086ms
2021-12-15T21:38:28.672342860Z panic: runtime error: invalid memory address or nil pointer dereference
2021-12-15T21:38:28.672380369Z [signal SIGSEGV: segmentation violation code=0x1 addr=0x120 pc=0xadf346]
2021-12-15T21:38:28.672384695Z 
2021-12-15T21:38:28.672389397Z goroutine 176 [running]:
2021-12-15T21:38:28.672393693Z github.com/celo-org/celo-blockchain/core/state.(*StateDB).Snapshot(0x0, 0x0)
2021-12-15T21:38:28.672398171Z 	github.com/celo-org/celo-blockchain/core/state/statedb.go:748 +0x26
2021-12-15T21:38:28.672402860Z github.com/celo-org/celo-blockchain/core/vm.(*EVM).StaticCall(0xc031d76000, 0x1998aa0, 0xc0317fe978, 0x0, 0x0, 0x10ce0000, 0xc031d40640, 0x24, 0x40, 0x186a0, ...)
2021-12-15T21:38:28.672407795Z 	github.com/celo-org/celo-blockchain/core/vm/evm.go:371 +0x82
2021-12-15T21:38:28.672415291Z github.com/celo-org/celo-blockchain/core/vm/vmcontext.(*evmRunner).Query(0xc03178bee0, 0x0, 0x0, 0x10ce0000, 0xc031d40640, 0x24, 0x40, 0x186a0, 0x0, 0x0, ...)
2021-12-15T21:38:28.672447804Z 	github.com/celo-org/celo-blockchain/core/vm/vmcontext/runner.go:74 +0x125
2021-12-15T21:38:28.672461769Z github.com/celo-org/celo-blockchain/contracts.(*BoundMethod).run(0xc00034a900, 0x19b6f58, 0xc03178bee0, 0x1746a60, 0xc0317fe948, 0x1, 0x0, 0xc031794b10, 0x1, 0x1, ...)
2021-12-15T21:38:28.672480126Z 	github.com/celo-org/celo-blockchain/contracts/contract_call.go:103 +0x3d3
2021-12-15T21:38:28.672513384Z github.com/celo-org/celo-blockchain/contracts.(*BoundMethod).Query(...)
2021-12-15T21:38:28.672518522Z 	github.com/celo-org/celo-blockchain/contracts/contract_call.go:76
2021-12-15T21:38:28.672521996Z github.com/celo-org/celo-blockchain/contracts.GetRegisteredAddress(0x19b6f58, 0xc03178bee0, 0x734f68e137d8918b, 0x94082397613bd753, 0x28722f2869f83c24, 0x373f3041b496df41, 0x0, 0x0, 0x0, 0x0, ...)
2021-12-15T21:38:28.672526277Z 	github.com/celo-org/celo-blockchain/contracts/registry.go:23 +0x1c6
2021-12-15T21:38:28.672566102Z github.com/celo-org/celo-blockchain/contracts.resolveAddressForCall(0x19b6f58, 0xc03178bee0, 0x734f68e137d8918b, 0x94082397613bd753, 0x28722f2869f83c24, 0x373f3041b496df41, 0x17c8c67, 0xd, 0x0, 0x0, ...)
2021-12-15T21:38:28.672571716Z 	github.com/celo-org/celo-blockchain/contracts/utils.go:41 +0x85
2021-12-15T21:38:28.672575385Z github.com/celo-org/celo-blockchain/contracts.NewRegisteredContractMethod.func1(0x19b6f58, 0xc03178bee0, 0x0, 0x0, 0xe00000000, 0xc00400bbd8, 0xc02de74f48)
2021-12-15T21:38:28.672603608Z 	github.com/celo-org/celo-blockchain/contracts/contract_call.go:60 +0xa8
2021-12-15T21:38:28.672632113Z github.com/celo-org/celo-blockchain/contracts.(*BoundMethod).run(0xc00034b020, 0x19b6f58, 0xc03178bee0, 0x156f3a0, 0xc02be300b0, 0x171e301, 0x0, 0x0, 0x0, 0x0, ...)
2021-12-15T21:38:28.672637031Z 	github.com/celo-org/celo-blockchain/contracts/contract_call.go:88 +0xb1
2021-12-15T21:38:28.672640928Z github.com/celo-org/celo-blockchain/contracts.(*BoundMethod).Query(...)
2021-12-15T21:38:28.672679854Z 	github.com/celo-org/celo-blockchain/contracts/contract_call.go:76
2021-12-15T21:38:28.672700787Z github.com/celo-org/celo-blockchain/contracts/blockchain_parameters.getBlockGasLimit(0x19b6f58, 0xc03178bee0, 0x15b7160, 0x6ee54b7ac4d5ae01, 0xc03129f338)
2021-12-15T21:38:28.672705351Z 	github.com/celo-org/celo-blockchain/contracts/blockchain_parameters/blockchain_parameters.go:93 +0x85
2021-12-15T21:38:28.672723072Z github.com/celo-org/celo-blockchain/contracts/blockchain_parameters.GetBlockGasLimitOrDefault(0x19b6f58, 0xc03178bee0, 0x8)
2021-12-15T21:38:28.672727872Z 	github.com/celo-org/celo-blockchain/contracts/blockchain_parameters/blockchain_parameters.go:82 +0x39
2021-12-15T21:38:28.672740787Z github.com/celo-org/celo-blockchain/ethstats.(*Service).assembleBlockStats(0xc0002fc000, 0xc008910200, 0x80)
2021-12-15T21:38:28.672800570Z 	github.com/celo-org/celo-blockchain/ethstats/ethstats.go:1027 +0x554
2021-12-15T21:38:28.672809454Z github.com/celo-org/celo-blockchain/ethstats.(*Service).reportBlock(0xc0002fc000, 0xc02c78c2e8, 0xc008910200, 0x0, 0x6)
2021-12-15T21:38:28.672813576Z 	github.com/celo-org/celo-blockchain/ethstats/ethstats.go:966 +0x4d
2021-12-15T21:38:28.672827340Z github.com/celo-org/celo-blockchain/ethstats.(*Service).loop.func5(0x0, 0x0)
2021-12-15T21:38:28.672868520Z 	github.com/celo-org/celo-blockchain/ethstats/ethstats.go:383 +0xf7a
2021-12-15T21:38:28.672873527Z golang.org/x/sync/errgroup.(*Group).Go.func1(0xc000d679e0, 0xc0004f6c80)
2021-12-15T21:38:28.672876821Z 	golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x59
2021-12-15T21:38:28.672880311Z created by golang.org/x/sync/errgroup.(*Group).Go
2021-12-15T21:38:28.672883636Z 	golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:54 +0x66
```

This PR fixes this issue by avoiding reporting stats that require access to the chain state when it
is unavailable. It also logs a warning message to indicate when this error occurs.

_Brief explanation of why these changes are/are not backwards compatible._

### CODE REVIEW

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit 8c883d47496418ff4a590af7e5163ad9aa6626d3
<details><summary>coverage:  45.3% of statements across all listed packages</summary><br>
<pre>
coverage:  54.6% of statements in consensus/istanbul
coverage:  45.2% of statements in consensus/istanbul/backend
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  43.3% of statements in consensus/istanbul/backend/internal/db
coverage:  24.1% of statements in consensus/istanbul/backend/internal/enodes
coverage:  22.4% of statements in consensus/istanbul/backend/internal/replica
coverage:  65.8% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
</pre>
CommentID: c0eae49909
</details>

**Review by gastonponti — COMMENTED** [review]
Nice catch!

I think you missed an else

File: ethstats/ethstats.go [hunk 1]
```diff
@@ -1019,12 +1015,19 @@ func (s *Service) assembleBlockStats(block *types.Block) *blockStats {
...
 
-	gasLimit := blockchain_parameters.GetBlockGasLimitOrDefault(vmRunner)
+		// only assemble every valSetInterval blocks
+		if block != nil && block.Number().Uint64()%valSetInterval == 0 {
+			valSet = s.assembleValidatorSet(block, stateDB)
+		}
+
+		vmRunner := s.backend.NewEVMRunner(header, stateDB)
+		gasLimit = blockchain_parameters.GetBlockGasLimitOrDefault(vmRunner)
+	}
```
[the remark this answers is not in the document]
Comment (nategraf) [reply ethstats/ethstats.go:1030]: 🤦

File: ethstats/ethstats.go [hunk 2]
```diff
@@ -1019,12 +1015,19 @@ func (s *Service) assembleBlockStats(block *types.Block) *blockStats {
...
-	if block != nil && block.Number().Uint64()%valSetInterval == 0 {
-		valSet = s.assembleValidatorSet(block, stateDB)
-	}
+	stateDB, _, err := s.backend.StateAndHeaderByNumberOrHash(context.Background(), rpc.BlockNumberOrHashWithHash(header.Hash(), true))
+
+	if err != nil {
+		log.Warn("Block state unavailable for reporting block stats", "hash", header.Hash(), "number", header.Number.Uint64(), "err", err)
 
-	gasLimit := blockchain_parameters.GetBlockGasLimitOrDefault(vmRunner)
+		// only assemble every valSetInterval blocks
```
[the remark this answers is not in the document]
Comment (nategraf) [reply ethstats/ethstats.go:1023]: 🤦

File: ethstats/ethstats.go [hunk 3]
```diff
@@ -1019,12 +1015,19 @@ func (s *Service) assembleBlockStats(block *types.Block) *blockStats {
...
 	blockRemain := epochSize - istanbul.GetNumberWithinEpoch(header.Number.Uint64(), epochSize)
 
-	// only assemble every valSetInterval blocks
-	if block != nil && block.Number().Uint64()%valSetInterval == 0 {
-		valSet = s.assembleValidatorSet(block, stateDB)
-	}
+	stateDB, _, err := s.backend.StateAndHeaderByNumberOrHash(context.Background(), rpc.BlockNumberOrHashWithHash(header.Hash(), true))
+
+	if err == nil {
+		log.Warn("Block state unavailable for reporting block stats", "hash", header.Hash(), "number", header.Number.Uint64(), "err", err)
```
Comment (piersy) [inline ethstats/ethstats.go:1021]: It seems like its not your day today @nategraf :D we want to log in the case of an error!

```suggestion
	if err != nil {
		log.Warn("Block state unavailable for reporting block stats", "hash", header.Hash(), "number", header.Number.Uint64(), "err", err)
	} else {
```

Comment (nategraf) [reply ethstats/ethstats.go]: 🙏  I guess I knew I was tired, but someone might need to take aware my dev license, haha. This coding is reckless endangerment.

Comment (piersy) [reply ethstats/ethstats.go]: hehe!

File: ethstats/ethstats.go [hunk 4]
```diff
@@ -981,11 +981,9 @@ func (s *Service) assembleBlockStats(block *types.Block) *blockStats {
 	// Gather the block infos from the local blockchain
 	var (
 		header   *types.Header
-		stateDB  *state.StateDB
-		vmRunner vm.EVMRunner
-		td       *big.Int
```
Comment (piersy) [inline ethstats/ethstats.go:986]: Just wondering why you moved td? Seems like it doesn't change the functionality but more likely to cause a conflict in a future merge.

Comment (gastonponti) [reply ethstats/ethstats.go:986]: The vmRunner is ours. Don't really see this possible conflict as an issue, haha

**Review by gastonponti — APPROVED** [review]
LGTM

File: ethstats/ethstats.go (hunk 4 above)
Comment (nategraf) [reply ethstats/ethstats.go:986]: Good call. I reverted the change to `td`. The others I don't think should introduce any conflict.

Comment (piersy) [reply ethstats/ethstats.go:986]: Thanks, moving vm and stateDB I think will reduce conflicts, since now that var block perfectly matches upstream

**Review by piersy — APPROVED** [review]
Looking good now, thanks Victor!

**Pull request merged**